### ISSUE TITLE
Metrics logged by self.log and metric.compute() are different

### ISSUE THREAD

**Issue by mfoglio** [Lightning-AI/pytorch-lightning#19816]
### Bug description

When testing my model, I am logging my F-1 metric every step using `self.log('f1_metric', self.f1_metric.compute(), on_epoch=True, on_step=False)`. 

At the end of the epoch, I found that `self.trainer.logged_metrics['f1_metric']` and `self.f1_metric.compute()` have two different values. This seems unexpected to me, as I never reset the metric value during the epoch.

If instead I log the metric using  `self.log('f1_metric', self.f1_metric.compute(), on_epoch=False, on_step=True)` the two printed values are the same. However, this is unexpected as I expected the `F1Score` instance to accumulate the value over the epoch as I never reset its state.

I am not sure if this can be considered a bug, but I still find it strange. What am I missing?

EDIT: if I use a batch_size bigger than the dataset, all the metrics logged have the same value. This seems to prove that it's an (expected?) issue with how `self.log` accumulates and reduces the value.

EDIT2: if I downgrade both PyTorch and torch lightning I get the same results for all the 3 metrics even when using `on_epoch=True` and `on_step=False` which is what is confusing me as I was originally working with these older versions of the libraries:
```
pytorch-lightning==1.0.0
pytorch-lightning-bolts==0.3.2.post1
sagemaker-pytorch-training==2.8.1
torch==1.10.1
torchmetrics==0.10.0
torchvision==0.17.0
```

### What version are you seeing the problem on?

v2.2

### How to reproduce the bug

```python
import torch
from torch import nn
import torch.nn.functional as F
from torch.utils.data import DataLoader, TensorDataset
import pytorch_lightning as pl
from pytorch_lightning import Trainer
import torchmetrics


class DummyModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        # Layers
        self.layer1 = nn.Linear(10, 64)  # Assume input features are of size 10
        self.layer2 = nn.Linear(64, 3)   # Output for 3 classes
        # F-1 Metric
        self.f1_metric = torchmetrics.F1Score(
            task='multiclass',
            ignore_index=-1,
            average='macro',
            num_classes=3,
            top_k=1
        )

    def forward(self, x):
        x = F.relu(self.layer1(x))
        x = self.layer2(x)
        return x
    
    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=0.001)
        return optimizer

    def test_step(self, batch, batch_id):
        x, y = batch
        logits = self(x)
        loss = F.cross_entropy(logits, y)
        self.f1_metric.update(logits, y)
        self.log('f1_metric', self.f1_metric.compute(), on_epoch=True, on_step=False)

    def on_test_epoch_end(self):
        print()
        print('Value logged:', self.trainer.logged_metrics['f1_metric'])
        print('Value computed: ', self.f1_metric.compute())
        self.f1_metric.reset()


# Dummy dataset
x_dummy = torch.randn(100, 10)  # 100 samples, 10 features each
y_dummy = torch.randint(0, 3, (100,))  # Random target labels for 3 classes

dataset = TensorDataset(x_dummy, y_dummy)
dataloader = DataLoader(dataset, batch_size=32, shuffle=True)

# Model
model = DummyModel()

# Trainer
trainer = Trainer(max_epochs=10, logger=True)
trainer.test(model, dataloader)
```

### Error messages and logs

```
Testing DataLoader 0: 100%|██████████| 50/50 [00:00<00:00, 51.39it/s]
Value logged: tensor(0.2091, device='mps:0')
Value computed:  tensor(0.2682, device='mps:0')
Testing DataLoader 0: 100%|██████████| 50/50 [00:01<00:00, 49.37it/s]
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         f1_metric         │    0.20905724167823792    │
└───────────────────────────┴───────────────────────────┘
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- OS (e.g., Linux): macOS
#- How you installed Lightning(`conda`, `pip`, source): pip
#- Running environment of LightningApp (e.g. local, cloud): local
```
requirements.txt
```
pytorch-lightning==2.2.0.post0
pytorch-lightning-bolts==0.3.2.post1
sagemaker-pytorch-training==2.8.1
torch==2.2.0
torchmetrics==0.10.0
torchvision==0.17.0
```

</details>

**Comment by mfoglio** [comment]
If I use torch lightning 1.0.1, I get the following output. Note that the first metric is different from the last 3:
```
Testing: 0it [00:00, ?it/s]--------------------------------------------------------------------------------
DATALOADER:0 TEST RESULTS
{'accuracy': tensor(0.4021)}
--------------------------------------------------------------------------------
Value logged: 0.36890244483947754
Value computed:  tensor(0.3689)
Accuracy computed from confusion matrix: 0.36890243902439024
Testing: 100%|██████████| 4/4 [00:00<00:00, 413.77it/s]
```
Starting from torch lightning 1.0.2, I get the following output where the first two metrics are equal but the last two are still different:
```
Testing: 0it [00:00, ?it/s]--------------------------------------------------------------------------------
DATALOADER:0 TEST RESULTS
{'accuracy': tensor(0.3884)}
--------------------------------------------------------------------------------
Value logged: 0.3884352743625641
Value computed:  tensor(0.3672)
Accuracy computed from confusion matrix: 0.36718564406191184
Testing: 100%|██████████| 4/4 [00:00<00:00, 663.92it/s]
```
@williamFalcon I know you worked on this a few years ago, but I saw your [release notes here ](https://github.com/Lightning-AI/pytorch-lightning/releases?page=13#:~:text=Compare-,fixes%20a%20major%20logging%20bug%20for%20val%20in%201.0,-Fixes%20the%20last) mentioning fixes related to validation metrics. Could you add more context to explain why the two metrics above and the ones below are different? 
Thank you!

**Comment by mfoglio** [comment]
I will keep adding my findings here. I am getting different results for `self.log` running the test multiple times on the same model and dataset. The only way I get reproducible results is if I use a batch size bigger than the dataset size.

_dataset.csv_ 
```
x,y
0,0
0,0
1,1
1,1
1,2
2,2
2,2
2,0
```
_Python code_
```
from io import BytesIO
from matplotlib import pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sn
import torch
from torch import nn
import torch.nn.functional as F
from torch.utils.data import DataLoader, TensorDataset
from PIL import Image
import pytorch_lightning as pl
from pytorch_lightning import Trainer
import torchmetrics
import torchvision


class Net(nn.Module):

    def __init__(self, n_output: int):
        super(Net, self).__init__()
        self.n_output = n_output

    def forward(self, batch_x):
        batch_y = torch.zeros((batch_x.shape[0], self.n_output), device=batch_x.device)
        for i, x in enumerate(batch_x):
            batch_y[i, x] = 1
        return batch_y


class DummyModel(pl.LightningModule):
    def __init__(self, n_output: int):
        super().__init__()
        # Layers
        self.net = Net(n_output)
        # F-1 Metric
        self.accuracy = torchmetrics.Accuracy(
            task='multiclass',
            ignore_index=-1,
            average='macro',
            num_classes=3,
            top_k=1
        )
        self.conf_matrix = torchmetrics.ConfusionMatrix(
            task='multiclass',
            ignore_index=-1,
            num_classes=3,
        )

    def forward(self, x):
        return self.net(x)

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=0.001)
        return optimizer

    def test_step(self, batch, batch_id):
        x, y = batch
        logits = self(x)
        loss = F.cross_entropy(logits, y)
        self.accuracy.update(logits, y)
        self.conf_matrix.update(logits, y)
        self.log('accuracy', self.accuracy.compute(), on_epoch=True, on_step=False)

    def on_test_epoch_end(self):
        accuracy_from_conf_matrix = self.compute_accuracy_from_confusion_matrix()

        print()
        print('Value logged:', self.trainer.logged_metrics['accuracy'])
        print('Value computed: ', self.accuracy.compute())
        print('Accuracy computed from confusion matrix:', accuracy_from_conf_matrix)

        self.accuracy.reset()

    def compute_accuracy_from_confusion_matrix(self):
        conf_mat = self.conf_matrix \
            .compute() \
            .detach() \
            .cpu() \
            .numpy()
        accuracies = list()
        for i in range(len(conf_mat)):
            accuracies.append(conf_mat[i, i] / np.sum(conf_mat[i]))
        accuracy = np.sum(accuracies) / len(accuracies)
        return accuracy


# Dummy dataset
df = pd.read_csv('dataset.csv')
x = torch.from_numpy(df['x'].to_numpy())
y = torch.from_numpy(df['y'].to_numpy())
dataset = TensorDataset(x, y)

# Model
model = DummyModel(n_output=len(y.unique()))

# Trainer
trainer = Trainer(max_epochs=10, logger=True, accelerator='cpu')

# Test
for batch_size in [1, 3, 10]:
    print(f'\nBatch size={batch_size}  (dataset size = {len(dataset)}')
    dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=False)
    trainer.test(model, dataloader)
```
Output
```
GPU available: True (mps), used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs

Batch size=1  (dataset size = 8
Testing DataLoader 0: 100%|██████████| 8/8 [00:00<00:00, 1541.17it/s]
Value logged: tensor(0.6458)
Value computed:  **tensor(0.7778)**
Accuracy computed from confusion matrix: 0.7777777777777777
Testing DataLoader 0: 100%|██████████| 8/8 [00:00<00:00, 1199.23it/s]
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         accuracy          │    **0.6458333134651184**     │
└───────────────────────────┴───────────────────────────┘

Batch size=3  (dataset size = 8
Testing DataLoader 0: 100%|██████████| 3/3 [00:00<00:00, 3728.27it/s]
Value logged: tensor(0.7569)
Value computed:  **tensor(0.7778)**
Accuracy computed from confusion matrix: 0.7777777777777777
Testing DataLoader 0: 100%|██████████| 3/3 [00:00<00:00, 2272.92it/s]
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         accuracy          │    **0.7569444179534912**     │
└───────────────────────────┴───────────────────────────┘

Batch size=10  (dataset size = 8
Testing DataLoader 0: 100%|██████████| 1/1 [00:00<00:00, 3775.25it/s]
Value logged: tensor(0.7778)
Value computed:  **tensor(0.7778)**
Accuracy computed from confusion matrix: 0.7777777777777777
Testing DataLoader 0: 100%|██████████| 1/1 [00:00<00:00, 1398.57it/s]
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         accuracy          │    **0.7777777910232544**     │
└───────────────────────────┴───────────────────────────┘
```

**Comment by mfoglio** [comment]
My bad. I should have used  `self.log('f1_metric', self.f1_metric, on_epoch=True, on_step=False)` instead of ` self.log('f1_metric', self.f1_metric.compute(), on_epoch=True, on_step=False)`. 
Hopefully, this post will be helpful to someone else!
